pkg/ec2pricing/ec2pricing.go

```diff
@@ -0,0 +1,312 @@
+)
+
+const (
+	defaultSpotDaysBack = 30
+	productDescription  = "Linux/UNIX (Amazon VPC)"
+	serviceCode         = "AmazonEC2"
+)
+
+// EC2Pricing is the public struct to interface with AWS pricing APIs
+type EC2Pricing struct {
+	PricingClient pricingiface.PricingAPI
+	EC2Client     ec2iface.EC2API
+	AWSSession    *session.Session
+	cache         map[string]float64
+	spotCache     map[string]map[string][]spotPricingEntry
+}
+
+// EC2PricingIface is the EC2Pricing interface mainly used to mock out ec2pricing during testing
+type EC2PricingIface interface {
+	GetOndemandInstanceTypeCost(instanceType string) (float64, error)
+	GetSpotInstanceTypeNDayAvgCost(instanceType string, availabilityZones []string, days int) (float64, error)
+	HydrateOndemandCache() error
+	HydrateSpotCache(days int) error
+}
+
+type spotPricingEntry struct {
+	Timestamp time.Time
+	SpotPrice float64
+}
+
+// New creates an instance of instance-selector EC2Pricing
+func New(sess *session.Session) *EC2Pricing {
+	return &EC2Pricing{
+		PricingClient: pricing.New(sess),
+		EC2Client:     ec2.New(sess),
+		AWSSession:    sess,
+	}
+}
+
+// GetSpotInstanceTypeNDayAvgCost retrieves the spot price history for a given AZ from the past N days and averages the price
+// Passing an empty list for availabilityZones will retrieve avg cost for all AZs in the current AWSSession's region
+func (p *EC2Pricing) GetSpotInstanceTypeNDayAvgCost(instanceType string, availabilityZones []string, days int) (float64, error) {
+	endTime := time.Now().UTC()
+	startTime := endTime.Add(time.Hour * time.Duration(24*-1*days))
+
+	spotPriceHistInput := ec2.DescribeSpotPriceHistoryInput{
+		ProductDescriptions: []*string{aws.String(productDescription)},
+		StartTime:           &startTime,
+		EndTime:             &endTime,
+		InstanceTypes:       []*string{&instanceType},
+	}
+	zoneToPriceEntries := map[string][]spotPricingEntry{}
+
+	if _, ok := p.spotCache[instanceType]; !ok {
+		var processingErr error
+		err := p.EC2Client.DescribeSpotPriceHistoryPages(&spotPriceHistInput, func(dspho *ec2.DescribeSpotPriceHistoryOutput, b bool) bool {
+			for _, history := range dspho.SpotPriceHistory {
+				var spotPrice float64
+				spotPrice, processingErr = strconv.ParseFloat(*history.SpotPrice, 64)
+				zone := *history.AvailabilityZone
+
+				zoneToPriceEntries[zone] = append(zoneToPriceEntries[zone], spotPricingEntry{
+					Timestamp: *history.Timestamp,
+					SpotPrice: spotPrice,
+				})
+			}
+			return true
+		})
+		if err != nil {
+			return float64(0), err
+		}
+		if processingErr != nil {
+			return float64(0), processingErr
+		}
+	} else {
+		for zone, priceEntries := range p.spotCache[instanceType] {
+			for _, entry := range priceEntries {
+				zoneToPriceEntries[zone] = append(zoneToPriceEntries[zone], spotPricingEntry{
+					Timestamp: entry.Timestamp,
+					SpotPrice: entry.SpotPrice,
+				})
+			}
+		}
+	}
+
+	aggregateZoneSum := float64(0)
+	numOfZones := 0
```

Comment:
it would work the same way, why do you ask?

Reply:
I updated `aggregateZoneSum` to `aggregateZonePriceSum` to differentiate what those two vars are used for since that was a little confusing when I was reading the code back.

Reply:
nothing wrong with the code just saw an inconsistency casting `float64()` in the return type vs. initialization. I think the next func down also returns an avg but initializes both as a float64 to start. not worth a revision :D